File: app/frontend/src/api/projects.ts
// 项目相关 API hooks。端点契约对齐 backend api/projects.py(M1-7 commit 2a189d1)+
// chapters.py(M1-9 commit 44e974c)。
//
// 端点:
//   - GET    /api/projects                       list[Project]
//   - POST   /api/projects                       Project(201)
//   - GET    /api/projects/{id}                  Project
//   - DELETE /api/projects/{id}                  {ok: true}
//   - POST   /api/projects/{id}/start            {run_id, queued}
//   - GET    /api/projects/{id}/documents        list[Document](M1+ commit 73d51ec)
//   - POST   /api/projects/{id}/documents        Document(multipart kind=tech_spec/scoring/template + file)
//   - GET    /api/projects/{id}/outline          OutlineResponse(含 chapters[] 完整状态)
//   - PUT    /api/projects/{id}/outline          {ok: true}
//   - GET    /api/projects/{id}/proposal         ProposalResponse
//   - GET    /api/projects/{id}/proposal.md      file response
//
// 注意:后端无独立 GET /chapters 列表;chapter 列表来自 /outline.chapters。
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { apiFetch } from '@/lib/apiFetch'
import type {
  DocumentDTO,
  DocumentKind,
  OutlineChapterIn,
  OutlineResponseDTO,
  ProjectDTO,
  ProposalResponseDTO,
  StartResponseDTO,
} from '@/lib/types'

const PROJECTS_KEY = ['projects'] as const

// 项目处于运行 / 等待用户决策态时需要前端持续轮询;命中即让 useProjects /
// useProject / useProjectOutline 共用同一组 polling 触发条件,避免「列表里
// 显示 awaiting_review 但详情页一直停在 extracting」之类的漂移。
const ACTIVE_PROJECT_STATUSES = new Set<string>([
  'queued',
  'extracting',
  'awaiting_material_understanding',
  'outlining',
  'outline_ready',
  'running',
  'awaiting_review',
])

export function useProjects() {
  return useQuery({
    queryKey: PROJECTS_KEY,
    queryFn: () => apiFetch<ProjectDTO[]>('/api/projects'),
    // R-12 全局进度感知:有 in-flight 项目时 5s 一次轮询,让 GlobalProgressBanner
    // 跨页能看到 status 变迁。无 in-flight 时停止 polling,避免空转。
    refetchInterval: (query) => {
      const data = query.state.data
      const hasActive = (data ?? []).some((p) =>
        ACTIVE_PROJECT_STATUSES.has(p.status),
      )
      return hasActive ? 5_000 : false
    },
  })
}

export function useProject(projectId: number | null) {
  return useQuery({
    queryKey: ['projects', projectId],
    queryFn: () => apiFetch<ProjectDTO>(`/api/projects/${projectId}`),
    enabled: projectId != null,
    // 项目处运行 / 等待态时 5s 轮询。`OutlineConfirmPage` /
    // `MaterialUnderstandingPage` / `DocumentUploadPage` 都以 useProject 派生
    // 按钮可用性;没有 polling 时拿到一次 extracting 后 LLM-0 完成 / 提纲
    // 完成都不会刷新,UI 卡死。
    refetchInterval: (query) => {
      const status = query.state.data?.status
      if (!status) return false
      return ACTIVE_PROJECT_STATUSES.has(status) ? 5_000 : false
    },
  })
}

export function useProjectOutline(projectId: number | null) {
  return useQuery({
    queryKey: ['projects', projectId, 'outline'],
    queryFn: () =>
      apiFetch<OutlineResponseDTO>(`/api/projects/${projectId}/outline`),
    enabled: projectId != null,
    // 提纲在 outlining → outline_ready → running 期间会从空 → 完整;章节
    // generating 时 partial final_text 也持续 flush。无 polling 时 P4/P5
    // 永远拿不到后续 chapter 状态。终态(done/failed/aborted)停。
    refetchInterval: (query) => {
      const status = query.state.data?.status
      if (!status) return 3_000 // 还没拿到 status,3s 重试
      return ACTIVE_PROJECT_STATUSES.has(status) ? 3_000 : false
    },
  })
}

// Phase 1C (2026-05-16):实体黑板 10 桶 JSON。404 = categorize_blackboard
// 还没跑或失败,前端展示「等 LLM 拆桶完成」占位。轮询逻辑:项目在
// extracting / outlining / queued 这几个还没看到桶的阶段每 3s 轮询,
// 有 payload 就停。
interface BlackboardEntitiesResponseDTO {
  project_id: number
  blackboard_entities: Record<string, BlackboardEntryDTO[]>
}
export interface BlackboardEntryDTO {
  tags: string[]
  content: string
  source_doc?: string
  section?: string
}
export function useBlackboardEntities(projectId: number | null) {
  return useQuery({
    queryKey: ['projects', projectId, 'blackboard-entities'],
    queryFn: () =>
      apiFetch<BlackboardEntitiesResponseDTO>(
        `/api/projects/${projectId}/blackboard-entities`,
      ),
    enabled: projectId != null,
    retry: false, // 404 是预期态(还没生成),不做指数退避
    refetchInterval: (q) => {
      if (q.state.data) return false
      const err = q.state.error
      if (!err) return false
      const msg = err instanceof Error ? err.message : String(err)
      // 只在「暂未就绪」的 404 上继续轮询;其它错(401/403/500)停
      return msg.includes('404') || msg.includes('暂未就绪')
        ? 3_000
        : false
    },
  })
}

export interface CreateProjectPayload {
  name: string
  description?: string | null
  pages_per_chapter?: number
  max_retry_per_chapter?: number
}

export function useCreateProject() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (body: CreateProjectPayload) =>
      apiFetch<ProjectDTO>('/api/projects', {
        method: 'POST',
        body: JSON.stringify(body),
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: PROJECTS_KEY })
    },
  })
}

export function useDeleteProject() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (projectId: number) =>
      apiFetch<{ ok: boolean }>(`/api/projects/${projectId}`, {
        method: 'DELETE',
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: PROJECTS_KEY })
    },
  })
}

export interface StartProjectPayload {
  pages_per_chapter?: number
  max_retry_per_chapter?: number
  outline_model?: string | null
  chapter_model?: string | null
  visuals_model?: string | null
}

export function useStartProject() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({
      projectId,
      body,
    }: {
      projectId: number
      body: StartProjectPayload
    }) =>
      apiFetch<StartResponseDTO>(`/api/projects/${projectId}/start`, {
        method: 'POST',
        body: JSON.stringify({
          pages_per_chapter: body.pages_per_chapter ?? 3,
          max_retry_per_chapter: body.max_retry_per_chapter ?? 3,
          outline_model: body.outline_model ?? null,
          chapter_model: body.chapter_model ?? null,
          visuals_model: body.visuals_model ?? null,
        }),
      }),
    onSuccess: (_data, vars) => {
      qc.invalidateQueries({ queryKey: ['projects', vars.projectId] })
      // 让全局列表 / banner 立刻看到 init → extracting 的状态切换,而不是
      // 等下一个 5s 周期
      qc.invalidateQueries({ queryKey: PROJECTS_KEY })
    },
  })
}

export function useProjectDocuments(projectId: number | null) {
  return useQuery({
    queryKey: ['projects', projectId, 'documents'],
    queryFn: () =>
      apiFetch<DocumentDTO[]>(`/api/projects/${projectId}/documents`),
    enabled: projectId != null,
    // 抽取是后台 arq 任务,.doc 走 LibreOffice 可能 5s+。任一文档 extract_status=pending
    // 时 2s 轮询,抽完后停止;让 DocumentUploadPage 的开始按钮能自动从 disabled 翻到可用。
    refetchInterval: (query) => {
      const data = query.state.data
      const hasPending = (data ?? []).some(
        (d) => d.extract_status === 'pending',
      )
      return hasPending ? 2_000 : false
    },
  })
}

export function useUploadDocument() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({
      projectId,
      kind,
      tags,
      file,
    }: {
      projectId: number
      // PR-M7-2:kind 可选,旧 UI 传值则保留;新 UI 用 tags
      kind?: DocumentKind | null
      tags?: string[]
      file: File
    }) => {
      const fd = new FormData()
      if (kind) fd.append('kind', kind)
      if (tags && tags.length > 0) fd.append('tags', tags.join(','))
      fd.append('file', file)
      return apiFetch<DocumentDTO>(`/api/projects/${projectId}/documents`, {
        method: 'POST',
        body: fd,
      })
    },
    onSuccess: (_data, vars) => {
      qc.invalidateQueries({
        queryKey: ['projects', vars.projectId, 'documents'],
      })
      qc.invalidateQueries({ queryKey: ['projects', vars.projectId] })
    },
  })
}

// PR-M7-2 多文件:支持删除单个上传文档,后端约束同 upload(项目仍处可编辑态)。
export function useDeleteDocument() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({
      projectId,
      documentId,
    }: {
      projectId: number
      documentId: number
    }) =>
      apiFetch<{ ok: boolean }>(
        `/api/projects/${projectId}/documents/${documentId}`,
        { method: 'DELETE' },
      ),
    onSuccess: (_data, vars) => {
      qc.invalidateQueries({
        queryKey: ['projects', vars.projectId, 'documents'],
      })
      qc.invalidateQueries({ queryKey: ['projects', vars.projectId] })
    },
  })
}

export interface ConfirmOutlinePayload {
  chapters: OutlineChapterIn[]
  // PR-M9-1:用户勾选的章节 id 列表。空 / null → 全选
  selected_chapter_ids?: string[] | null
  // textarea TOC + revise:默认 'confirm';'revise' 时只发 feedback
  decision?: 'confirm' | 'revise'
  feedback?: string | null
}

export function useConfirmOutline() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({
      projectId,
      chapters,
      selected_chapter_ids,
      decision,
      feedback,
    }: {
      projectId: number
      chapters: OutlineChapterIn[]
      selected_chapter_ids?: string[] | null
      decision?: 'confirm' | 'revise'
      feedback?: string | null
    }) =>
      apiFetch<{ ok: boolean }>(`/api/projects/${projectId}/outline`, {
        method: 'PUT',
        body: JSON.stringify({
          decision: decision ?? 'confirm',
          feedback: feedback ?? null,
          chapters,
          selected_chapter_ids: selected_chapter_ids ?? null,
        }),
      }),
    onSuccess: (_data, vars) => {
      qc.invalidateQueries({ queryKey: ['projects', vars.projectId] })
      // outline_ready → outlining(revise)/ running(confirm)的过渡也要让
      // 列表 banner 立刻刷一次
      qc.invalidateQueries({ queryKey: PROJECTS_KEY })
    },
  })
}

export function useProposalMarkdown(projectId: number | null) {
  return useQuery({
    queryKey: ['projects', projectId, 'proposal'],
    queryFn: () =>
      apiFetch<ProposalResponseDTO>(`/api/projects/${projectId}/proposal`),
    enabled: projectId != null,
  })
}

File: app/backend/src/bid_app/schemas/projects.py
"""Pydantic v2 IO schemas — projects 相关。"""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

# ⭐ M1 contract:Project.status 全集(单一来源,API 收紧用)
# init / queued:刚建项目还没 /start
# extracting / outlining:/start 后 LangGraph 阶段
# awaiting_material_understanding (PR-M8-1):material_understanding_review 节点
#   interrupt,等用户在 MaterialUnderstandingPage 上点 pass / revise / skip
# outline_ready:LLM-1 跑完,等用户在 P4 编辑提纲并 /confirm-outline
# running / awaiting_review:章节循环阶段(P5)
# done / failed / aborted:终态
# aborted_v1 (PR-M7-1):v2 上线时 ``flush_running_workflows`` CLI 把所有
#   in-flight v1 项目标记为该状态,提示用户重建。
# aborted_schema_v1 (PR-M7-1):worker 运行时检测到 ``WorkflowSchemaMismatch``
#   (老 checkpoint + 新 graph)时自动标记。
ProjectStatus = Literal[
    "init",
    "queued",
    "extracting",
    "awaiting_material_understanding",
    "outlining",
    "outline_ready",
    "running",
    "awaiting_review",
    "done",
    "failed",
    "aborted",
    "aborted_v1",
    "aborted_schema_v1",
]

# Chapter.status 已在 schemas/chapters.py 收紧;outline 端点回填 chapter 嵌入信息
# 时这里复用一份(避免循环 import)
ChapterStatusEnum = Literal[
    "pending",
    "generating",
    "awaiting_review",
    "reviewing",
    "approved",
    "skipped",
    "failed",
    "retrying",
]


class ProjectCreateRequest(BaseModel):
    name: str = Field(..., max_length=255)
    description: str | None = Field(None, max_length=2000)
    pages_per_chapter: int = Field(3, ge=1, le=10)
    max_retry_per_chapter: int = Field(3, ge=0, le=10)


class ProjectResponse(BaseModel):
    """⭐ M1 任务 #4:**不**暴露 ``dir_path`` / ``api_key_owner`` 给前端。

    - ``dir_path``:服务器端文件路径(``/data/projects/<id>`` 之类),纯
      实现细节,前端拿了也用不上,反而泄露内部目录结构(让外部扫描者
      推断容器内 layout)。需要内部用(下载 / DOCX 路径拼接)的代码继续
      从 ORM ``Project.dir_path`` 取,**不**经过 schema。
    - ``api_key_owner``:用户内部 ID。SSE 团队共享池设计下,其他成员
      看见"项目正在跑"就够了,具体是谁的 API key 在掏配额是结算 / 后台
      统计层面的事(/api/admin),不应在普通 P2 列表 / P3 详情 expose。
    """

    id: int
    name: str
    description: str | None
    status: ProjectStatus
    created_by: int
    # JOIN users.username 后填入;User 被删的极端情况留 None。
    created_by_username: str | None = None
    pages_per_chapter: int
    max_retry_per_chapter: int
    created_at: datetime

    model_config = {"from_attributes": True}


class StartRequest(BaseModel):
    pages_per_chapter: int = Field(3, ge=1, le=10)
    max_retry_per_chapter: int = Field(3, ge=0, le=10)
    outline_model: str | None = Field(None, max_length=128)
    chapter_model: str | None = Field(None, max_length=128)
    visuals_model: str | None = Field(None, max_length=128)


class StartResponse(BaseModel):
    run_id: int
    queued: bool


class OutlineChapterIn(BaseModel):
    """提纲编辑端的章节(用户改过的字段)。"""

    id: str | None = None
    # PR-M8-2 follow-up:层级编号 "1.1" / "2.3.1";None 时后端按 index+1 兜底
    section: str | None = None
    title: str = Field(..., min_length=1)
    summary: str | None = None
    key_points: list[str] = Field(..., min_length=1)
    target_pages: int = Field(..., ge=1, le=10)
    matched_scoring_items: list[str] = Field(default_factory=list)
    chapter_model: str | None = Field(None, max_length=128)


class OutlineConfirmRequest(BaseModel):
    """body.chapters 为空数组 / 缺失 → 自动确认沿用 LLM-1。

    PR-M9-1:可选 ``selected_chapter_ids`` 让前端在锁定目录时一并提交
    用户勾选的章节;空 / None → 全选(向后兼容)。

    textarea TOC + revise(PR-M8-2 follow-up #2):
    - ``decision="confirm"``(默认):提交 ``chapters`` 锁定目录
    - ``decision="revise"``:把 ``feedback`` 送 LLM-1 重新生成大纲
    """

    decision: Literal["confirm", "revise"] = "confirm"
    feedback: str | None = None
    chapters: list[OutlineChapterIn] = Field(default_factory=list)
    selected_chapter_ids: list[str] | None = Field(default=None)


class OutlineChapterOut(BaseModel):
    id: str
    title: str
    summary: str | None
    key_points: list[str]
    target_pages: int


class OutlineChapterDTO(BaseModel):
    """⭐ M1 contract:GET /outline 返的 chapter 嵌入对象。

    旧 schema 是 ``dict[str, Any]``,前端只能字符串字面量切字段;现在收紧
    成 Pydantic model,FastAPI 自动 OpenAPI 暴露字段名 + 类型。
    """

    id: str
    # PR-M8-2 follow-up:层级目录编号 "1.1" / "2.3.1";老项目可能 None
    section: str | None = None
    title: str
    summary: str | None = None
    key_points: list[str] = Field(default_factory=list)
    target_pages: int
    index: int
    status: ChapterStatusEnum
    chapter_model: str | None = None
    # ⭐ R-15 配套:R-14 partial / 完整正文从 outline 端点暴露给前端 hydrate
    final_text: str | None = None


class OutlineResponse(BaseModel):
    project_id: int
    run_id: int | None
    status: ProjectStatus
    max_concurrent_chapter_generations: int
    chapters: list[OutlineChapterDTO]


class DocumentUploadResponse(BaseModel):
    id: int
    project_id: int
    # PR-M7-2:kind 不再强制三选一;tags 取代分类语义。
    kind: str | None = None
    original_filename: str
    file_size: int
    byte_size: int | None = None
    mime_type: str | None = None
    tags: list[str] | None = None
    extract_error: str | None = None
    # PR-M7-2:异步抽取状态。pending = 抽取 task 已入队但未跑完。
    extract_status: Literal["pending", "done", "failed"] = "pending"
